### ISSUE TITLE
More extensive testing of Insieme required?

### ISSUE THREAD

**Issue by philippgs** [insieme/insieme#38]
Do we need a new build type or some other facility that allows more extensive testing of Insieme which is only omitted due to performance reasons? For example, #37 mentions disabled backend assertions, and I remember semantic errors popping up for specific (large) problem classes of e.g. NAS BT that are usually not tested due to their execution time. Also, there are probably some unit and/or integration tests that are not run solely for this reason.

If we do, one possible way to go would be a Jenkins job that runs only e.g. once per week. In any case we'd need some mechanism for enabling both compiler features as well as unit or integration tests.

**Comment by 0ife** [comment]
More testing is always good, but hand-coding these tests is busywork and time-consuming. We should automate as much as possible. I would propose to have a look at a fuzzer, for example http://lcamtuf.coredump.cx/afl/
Ideally, we would integrate such a tool into our Jenkins infrastructure; I second a run every week.

**Comment by klois** [comment]
The question is if we want to make Insieme really as general as possible or if we focus on the actual usecases that we have. At least for C++ and OpenCL we will never support the full standard and I don't know how we could tell the fuzzer not to create test cases which we don't support and also don't need to support.

**Comment by 0ife** [comment]
If we do not support specific use cases, we should not crash - we should tell the user that a feature is not supported. We can simply make a difference with proper exit codes.

**Comment by philippgs** [comment]
The original motivation for NOT doing this was the fact that a code generator (which is what a fuzzer is in the world of compilers) would quite possibly generate code variants and combinations that admittedly would cause Insieme to crash, but would never appear in real world codes. Hence, we feared fixing bugs in Insieme that will otherwise never arise, and we chose to focus on the actual codes we're using.

Also, the original idea of this question is not to have additional tests that still need to be written, but rather to enable those already present but disabled for performance reasons.

I am not against using such a tool or method per se, but I doubt its applicability in our case and its main objective, security, does not exist in our case.
